app/tests/spec/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,179 @@
+const encryptionOptions = {
+  iv: 'eeddccbbaa99887766554433'
+};
+
+describe('lib/crypto/recovery-keys', () => {
+  describe('getRecoveryKeyVersion', () => {
+    it('should get current recovery key version', () => {
+      const version = RecoveryKeys.getCurrentRecoveryKeyVersion();
+      assert.equal(version, '1', 'current version matches');
+    });
+  });
+
+  describe('generateRecoveryKey', () => {
+    it('should generate base32 string', () => {
+      return RecoveryKeys.generateRecoveryKey(1000)
+        .then((key) => {
+          assert.ok(/[0-9A-Z]+$/.test(key), 'no lowercase letters');
+          assert.equal(key.indexOf('I'), -1, 'no I');
+          assert.equal(key.indexOf('L'), -1, 'no L');
+          assert.equal(key.indexOf('O'), -1, 'no O');
+          assert.equal(key.indexOf('U'), -1, 'no U');
+        });
+    });
+
+    it('should prepend version', () => {
+      return RecoveryKeys.generateRecoveryKey(1000)
+        .then((key) => assert.equal(key.charAt(0), RecoveryKeys.getCurrentRecoveryKeyVersion()));
+    });
+
+    it('should have correct length', () => {
+      return RecoveryKeys.generateRecoveryKey(1000)
+        .then((key) => assert.lengthOf(key, 1000));
+    });
+  });
+
+  describe('getRecoveryJwk', () => {
+    it('throws if no uid', () => {
+      return RecoveryKeys.getRecoveryJwk(undefined, recoveryKey)
+        .then(assert.fail, (err) => {
+          assert.equal(err.message, 'Missing uid');
+        });
+    });
+
+    it('throws if no recoveryKey', () => {
+      return RecoveryKeys.getRecoveryJwk(uid, undefined)
+        .then(assert.fail, (err) => {
+          assert.equal(err.message, 'Missing recoveryKey');
+        });
+    });
+
+    it('should create recoveryJwk', () => {
+      return RecoveryKeys.getRecoveryJwk(uid, recoveryKey)
+        .then((jwk) => {
+          assert.ok(jwk, 'jwk returned');
+          assert.equal(jwk.alg, 'A256GCM', 'correct algorithm');
+          assert.equal(jwk.kty, 'oct', 'correct kty');
+          assert.equal(jwk.kid, expectedRecoveryKeyId, 'correct kid');
+          assert.equal(jwk.length, 256, 'correct length');
+        });
+    });
+  });
+
+  describe('bundleRecoveryData', () => {
+    let recoveryJwk;
+
+    beforeEach(() => {
+      return RecoveryKeys.getRecoveryJwk(uid, recoveryKey)
+        .then((result) => {
+          recoveryJwk = result;
+        });
+    });
+
+    it('throws if no recoveryJwk', () => {
+      return RecoveryKeys.bundleRecoveryData(undefined, recoveryData)
+        .then(assert.fail, (err) => {
+          assert.equal(err.message, 'Missing recoveryJwk');
+        });
+    });
+
+    it('should create recovery bundle', () => {
+      return RecoveryKeys.bundleRecoveryData(recoveryJwk, recoveryData, encryptionOptions)
+        .then((bundle) => {
+          assert.ok(bundle, 'bundle exists');
+          const tokens = bundle.split('.');
+          // The 0,1 and 4th tokens of the bundle are different on each test run
```

Comment:
Are they actually different on each run, or do they just differ from the python-generated test vectors?  I would expect them to be consistent on each run when using the same IV.

Reply:
@vbudhram I figured out how to update my python test-vector script to match the way JSON is formatted here.  Latest version:

  https://gist.github.com/rfk/0462fc6c81e43f2b9abbea41c479a784

It gives the following results:

```
$ python recovery_key_test_vectors.py
uid = aaaaabbbbbcccccdddddeeeeefffff00
recovery_key = 0123456789ABCDEFGHJKMNPQRSTVWXYZ
recover-kid =  6aa248931704886f54ac64b81b111bc0
recover-enc =  58c77394e516799304a06fa370828a2570b285ea7e461ed4d792830eeffe6d1a
kB = 000000111111222222333333444444555555666666777777888888999999ABCD
recover-data plaintext = {"kB":"000000111111222222333333444444555555666666777777888888999999ABCD"}
IV = eeddccbbaa99887766554433
recover-data = eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwia2lkIjoiNmFhMjQ4OTMxNzA0ODg2ZjU0YWM2NGI4MWIxMTFiYzAifQ..7t3Mu6qZiHdmVUQz.qoXWfU77FBV67Pyy_GXAkoSuAbksIqAXZd2tkq48iD3u_qWEEAeYx-M3m2G373zes3OXsSZdJIGRFmd8bC9s4ZvN3kF3A6vK5g.xMRqAVVJw-9MbolcZl4TZA
recover-data decrypts to: {"kB":"000000111111222222333333444444555555666666777777888888999999ABCD"}
```

If you fix the issue with using the hex-encoded keyMaterial rather than the raw bytes, I think the full `recover-data` JWE should come out identical to the value above, which will be a great testcase.